contracts/wallet/BaseWallet.sol

```diff
@@ -117,42 +120,45 @@ contract BaseWallet {
      * @param _data The data of the transaction.
      */
-    function invoke(address _target, uint _value, bytes calldata _data) external moduleOnly returns (bytes memory _result) {
+    function invoke(address _target, uint _value, bytes calldata _data) external virtual moduleOnly returns (bytes memory _result) {
         bool success;
         // solium-disable-next-line security/no-call-value
-        (success, _result) = _target.call.value(_value)(_data);
+        (success, _result) = _target.call{value: _value}(_data);
         if (!success) {
             // solium-disable-next-line security/no-inline-assembly
             assembly {
-                returndatacopy(0, 0, returndatasize)
-                revert(0, returndatasize)
+                returndatacopy(0, 0, returndatasize())
+                revert(0, returndatasize())
             }
         }
         emit Invoked(msg.sender, _target, _value, _data);
     }
 
     /**
-     * @dev This method makes it possible for the wallet to comply to interfaces expecting the wallet to
-     * implement specific static methods. It delegates the static call to a target contract if the data corresponds
-     * to an enabled method, or logs the call otherwise.
+     * @dev This method delegates the static call to a target contract if the data corresponds
+     * to an enabled module, or logs the call otherwise.
      */
-    function() external payable {
-        if (msg.data.length > 0) {
-            address module = enabled[msg.sig];
-            if (module == address(0)) {
-                emit Received(msg.value, msg.sender, msg.data);
-            } else {
-                require(authorised[module], "BW: must be an authorised module for static call");
-                // solium-disable-next-line security/no-inline-assembly
-                assembly {
-                    calldatacopy(0, 0, calldatasize())
-                    let result := staticcall(gas, module, 0, calldatasize(), 0, 0)
-                    returndatacopy(0, 0, returndatasize())
-                    switch result
-                    case 0 {revert(0, returndatasize())}
-                    default {return (0, returndatasize())}
-                }
+    fallback() external payable {
+        address module = enabled[msg.sig];
+        if (module == address(0)) {
+            emit Received(msg.value, msg.sender, msg.data);
+        } else {
+            require(authorised[module], "BW: must be an authorised module for static call");
+            // solium-disable-next-line security/no-inline-assembly
+            assembly {
+                calldatacopy(0, 0, calldatasize())
+                let result := staticcall(gas(), module, 0, calldatasize(), 0, 0)
+                returndatacopy(0, 0, returndatasize())
+                switch result
+                case 0 {revert(0, returndatasize())}
+                default {return (0, returndatasize())}
             }
         }
     }
+
+    receive() external payable {
```

Comment:
This function is not needed and can be removed entirely. It will never be reached from the Proxy (since calls to the Proxy with `msg.data == 0` are not forwarded by the Proxy).

Reply:
It's actually used in tests as we're not creating wallets that are Proxies but just new-ing up a `BaseWallet` instance. It's arguable whether this is correct and we shouldn't be testing with a proper Proxy-based wallet.

Reply:
It helps fix the coverage problem though as I can remove the logic of `receive` which allows coverage to pass.

Reply:
Yes but removing that method won't break the tests as they currently stand. The `fallback()` of BaseWallet will be executed instead, so I believe we can remove `receive()` regardless of whether we decide to rewrite the tests to use `Proxy` or not

Reply:
I does break the tests as fallback won't execute when there's no calldata received. That was the whole point of splitting the fallback function.

Reply:
As per the doc: 
```
A payable fallback function is also executed for plain Ether transfers, if no receive Ether function is present. [...]
```
https://solidity.readthedocs.io/en/v0.6.10/contracts.html#fallback-function

I also verified in Remix that the following contract is able to receive ether via empty calldata:
```
pragma solidity ^0.6.10;
contract FallbackTest {
    fallback() external payable {}
}
```

Reply:
Okay it's not failing because plain ether is not accepted but because the call to `fallback` costs more than 21000 gas. https://app.circleci.com/pipelines/github/argentlabs/argent-contracts/707/workflows/26010b94-04c4-4a0c-ad62-c9cbe0e39afe/jobs/897
I can't think of any way to completely remove the `receive` function off `BaseWallet` really but feel free to push any changes you think are possible.

Reply:
I see. I don't love the idea of having a `receive()` function in `BaseWallet` that is only there for tests. On the other hand having to use a Proxy + BaseWallet in all our tests feels a bit heavy. Maybe using a `BaseWalletTest` in our tests, that inherits from `BaseWallet` and only adds an empty `receive()` to `BaseWallet` could be a middle ground? If not, let's just keep the empty `receive()`.

Reply:
Thinking about this more, testing with a Proxy + BaseWallet is probably the right move. We could maybe add a `createWallet()` method to `test-manager.js` that does the deployment of both contracts + wrapping of the proxy with the BaseWallet abi.